listener/pkg/metrics/metrics.go

```diff
@@ -0,0 +1,85 @@
+package metrics
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	listenerRequestDuration            = "watcher_listener_request_duration"
+	listenerRequests                   = "watcher_listener_requests_total"
+	listenerRequestErrors              = "watcher_listener_request_errors_total"
+	listenerInflightRequests           = "watcher_listener_inflight_requests"
+	listenerExceedingSizeLimitRequests = "watcher_listener_exceeding_size_limit_requests_total"
+	listenerFailedVerificationRequests = "watcher_listener_failed_verification_requests_total"
+	requestURILabel                    = "request_uri_label"
+	listenerService                    = "listener"
+	serverNameLabel                    = "server_name"
+)
+
+var (
+	httpRequestDurationHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals
+		Name: listenerRequestDuration,
+		Help: "Indicates the latency of each request in seconds",
+	}, []string{serverNameLabel})
+	httpRequestsCounter = prometheus.NewCounterVec(prometheus.CounterOpts{ //nolint:gochecknoglobals
+		Name: listenerRequests,
+		Help: "Indicates the number of requests",
+	}, []string{serverNameLabel})
+	httpRequestErrorsCounter = prometheus.NewCounterVec(prometheus.CounterOpts{ //nolint:gochecknoglobals
+		Name: listenerRequestErrors,
+		Help: "Indicates the number of failed requests",
+	}, []string{serverNameLabel})
+	HTTPInflightRequestsGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals
+		Name: listenerInflightRequests,
+		Help: "Indicates the number of inflight requests",
+	}, []string{serverNameLabel})
+	httpRequestsExceedingSizeLimitCounter = prometheus.NewCounterVec(prometheus.CounterOpts{ //nolint:gochecknoglobals
+		Name: listenerExceedingSizeLimitRequests,
+		Help: "Indicates the number of requests exceeding size limit",
+	}, []string{serverNameLabel})
+	httpFailedVerificationRequests = prometheus.NewCounterVec(prometheus.CounterOpts{ //nolint:gochecknoglobals
+		Name: listenerFailedVerificationRequests,
+		Help: "Indicates the number of requests that failed verification",
+	}, []string{serverNameLabel, requestURILabel})
+)
+
+func InitMetrics(metricsRegistry metrics.RegistererGatherer) {
```

Comment:
you should not introduce `controller-runtime` as dependencies here as we discussed, instead, this should be `prometheus.Registry`

Reply:
I tried it with prometheus.Registry but that caused an error in the LCM because the ctrlMetrics.Registry is of the RegistererGatherer type that’s why I changed it. Another other option would be to register them in LCM, and the Init would be sth like Get and return an array of the counters/histogram,..